### PR TITLE
aotf: editable mutations

### PR DESCRIPTION
**Pull request by oliver-sanders**
Addresses #339 
Follow-on from: #504 
Companion Cylc Flow PR: https://github.com/cylc/cylc-flow/pull/3986 - merged

Example showing the trigger mutation being called, first using the configured defaults, then by edit.

aotf

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [ ] Appropriate tests are included (unit and/or functional).
- [ ] Appropriate change log entry included.
- [x] No documentation update required.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
Spotted two issues so far.

1. The broadcast mutation is getting a 400 return code.
    * Checked the mutation/args, looks good, not sure what's going on.
2. The workflow state can get stuck sometimes (not related to this PR).
   * Try holding and releasing it a few times, the play/pause button and GScan indicator should change.
   * However, sometimes they don't.
   * I have replicated this in GraphiQL so looks like this is Flow/UIS related.

**Comment by oliver-sanders** [comment]
Fixed the broadcast issue by treating broadcast settings in a specially, requires https://github.com/cylc/cylc-flow/pull/3986.

**Comment by oliver-sanders** [comment]
I've added some dummy mutations to the mocked workflow service so they can now be tested in offline mode and the first functional test, will test the mutation forms etc in due course.

**Comment by kinow** [comment]
@oliver-sanders I can write the tests for this one if you'd like? I just remembered how to use Cypress & write e2e tests working on #555 (slowly switching from vacation-mode to work-mode :smile:).

That way you can focus on other issues.

**Comment by oliver-sanders** [comment]
That's a very kind offer, thanks. Just realised I hadn't committed on this PR since November 🤦.

I had added some mock-mutations for the offline mode and had started writing tests for them but was running into issues running older Cypress on MacOS (had to quit and restart Cypress for every change which made the test development process painfully slow).

**Comment by kinow** [comment]
> That's a very kind offer, thanks. Just realised I hadn't committed on this PR since November facepalm.
> 
> I had added some mock-mutations for the offline mode and had started writing tests for them but was running into issues running older Cypress on MacOS (had to quit and restart Cypress for every change which made the test development process painfully slow).

Ah! So you should be able to finish the tests, hopefully without having to wait so much after #555 :crossed_fingers:

**Comment by oliver-sanders** [comment]
... one month later 🤦

The good news is that Cypress has worked like a charm since the upgrade, seriously impressive project 👏!

I've added functional tests to test that cylc-objects (cycle point, families, tasks, ...) are associated with mutations correctly.

Now I'm having a go at capturing `$workflowService.mutate` calls to test that the mutations are being executed correctly. I'm not quite sure how to mock the workflow service for this purpose, I've had a go at this:

```js
  it('fires the mutation when clicked', () => {
    cy.visit('/#/workflows/one')

    const mutations = []
    cy.window().its('app').then(app => {
      // Object.getPrototypeOf(app.$workflowService).mutate = (stuff) => {
      app.$workflowService.mutate = (stuff) => {
        mutations.push(stuff) // add the mutation arguments to an array for later comparison
        console.log('here')
      }
    })

    // open the mutations panel
    cy
      .get('.node-data-cyclepoint > .c-task:first')
      .should('exist')
      .should('be.visible')
      .click()

    // run the first mutation in the list
    cy
      .get('.c-mutation-menu-list:first')
      .find('.v-list-item__content:first')
      .should('exist')
      .should('be.visible')
      .click()
      .then(() => {
        // ensure the mutation would have been executed
        expect(mutations).to.deep.equal([])
      })
  })
```

No luck with that approach :(

**Comment by kinow** [comment]
Does it work?

```js
    cy.window().its('app.$workflowService').then(service => {
      // Object.getPrototypeOf(app.$workflowService).mutate = (stuff) => {
      cy.stub(service, 'mutate', () => {
        mutations.push(stuff) // add the mutation arguments to an array for later comparison
        console.log('here')
      })
    })
```

There's an example in the e2e test [`layout.js`](https://github.com/cylc/cylc-ui/blob/f9a4ac0c77a7cb2399081179f3ec5c18982a719c/tests/e2e/specs/layout.js#L25).

**Comment by oliver-sanders** [comment]
Ah, need to mock the `workflowService.apolloClient.mutate` method not `workflowService.mutate`...

**Comment by kinow** [comment]
> Ah, need to mock the `workflowService.apolloClient.mutate` method not `workflowService.mutate`...

Looks like you're almost done! It'll be great to have this PR merged. All mutations available from within the UI. I think you should be able to use `.stub` to modify the `apolloClient` too: https://docs.cypress.io/api/commands/stub.html#Syntax

**Comment by oliver-sanders** [comment]
Stub looks nice, thanks for the pointer, I think the `apolloClient` isn't defined so I couldn't easily use stub, however, the test is pretty simple so I was able to get away with just `service.apolloClient = { mutate () => {} }`.

Will come back to the unit tests tomorrow, hopefully get this finished soon 😓.

**Comment by kinow** [comment]
Looks like CI is ok now! :rocket:

**Comment by oliver-sanders** [comment]
Sorted out the play/pause and stop buttons in the toolbar!

The toolbar now sets an internal flag when it "mutates" a property and unsets that when the state changes. This property is then used to deactivate the buttons.


Note:

* This is a temporary solution until we have the ability to subscribe to mutation progress.
* Sometimes mutations fail or don't have the desired effect due to the timing of the request.
* So the buttons aren't deactivated, they are just greyed out, clicking them again will repeat the request.
* The stopping status doesn't seem to come through causing the play button to remain enabled whilst the workflow is supposed to be stopping.
* In the gif below I've added `<pre>{{ currentWorkflow.status }}</pre>`.

hold-stop

**Comment by kinow** [comment]
Half-way through the code. Then review with a couple running workflows. But should finish reviewing it this morning or this afternoon :tada: Looking great so far!

**Review by kinow — APPROVED** [review]
Few comments in the code, but no blockers. No need to update the code @oliver-sanders, mainly comments here and there. Tested UI + UIServer with couple workflows. Worked like a charm! Set outputs, release, hold, kill tasks. Also play/pause/stop workflow. :clap: and great tests!

I think for later we probably want a confirmation message before stopping workflows. WDYT @oliver-sanders @hjoliver ? I accidentally stopped my workflow during the tests :smile:

File: src/utils/aotf.js [hunk 1]
```diff
@@ -368,23 +368,13 @@ export function getIntrospectionQuery () {
...
- *     {
- *       args: [
- *         {
- *           _cylcObject: string,
- *           _required: boolean
- *         }
- *       ]
- *     }
+ *     },
+ *     requiresInfo: boolean
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/utils/aotf.js:372]: I'm not really familiar with JSDoc, happy to reformat it, however, I'm not sure how to do it. Like this?

```js
/*
 * Returns something with the shape:
 * 
 * {[
 *   [
 *     {
 *       args: [
 *         {
 *           _cylcObject: string,
 *           _required: boolean
 *         }
 *       ]
 *     },
 *     requiresInfo: boolean
 *   ]
 * ]} [satisfied, all]
 *
 * @param {} ...
 * ...
 * @returns {[*]} - mutations
 */
```

File: src/components/graphqlFormGenerator/components/BroadcastSetting.vue [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    },
+
+    fromString (string_) {
+      const [lhs, rhs] = this.lsplit(string_, '=', 2)
+      if (rhs === undefined) {
+        return
+      }
+      // const [lhs, rhs] = JavaSplit(string_, '=', 2)
+      if (lhs === 'inherit') {
+        return
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply src/components/graphqlFormGenerator/components/BroadcastSetting.vue:49]: :+1: Ah, I didn't realise that a bare `return` produced an `undefined`.

**Comment by oliver-sanders** [comment]
Thanks @kinow, good suggestions.

Prompting on actions is something we need to look at, for example the `cylc clean` command should definitely prompt.

In Cylc8 the "stop" button is now safe 🥳, you can always press "play" and everything will return to the way it was, so the need for a prompt on the stop button is much less than it was before. Trying to remember the Cylc7 GUI, I don't think we prompt for the stop button (but do for control => stop, which is equivalent to the new mutation editor).

File: src/utils/aotf.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (kinow) [reply src/utils/aotf.js:372]: That would be valid, but wouldn't help someone using JSdoc or an IDE. The function is returning an array with two values that I think the client/user expects to be in the order `Object, boolean`. I think in Python we can document that, but not with JS :disappointed_relieved: 

Here's the valid annotation for the current return type:

```js
 * @returns {[
 *   [
 *     {
 *       args: [
 *         {
 *           _cylcObject: string,
 *           _required: boolean
 *         }
 *       ]
 *     }|boolean
 *   ]
 * ]} [satisfied, all]
```

Basically same as `Union[object, bool]` in Py I think? But there's no guarantee that the first element of the array is an Object, and the second a boolean. Only guarantees that the array has only these two types.

When I need to return something like that in JS, I normally use an array of objects. Like this:

```diff
diff --git a/src/utils/aotf.js b/src/utils/aotf.js
index 45597aea..c61ceccc 100644
--- a/src/utils/aotf.js
+++ b/src/utils/aotf.js
@@ -360,8 +360,8 @@ export function getIntrospectionQuery () {
  * ]} mutations - Array of mutations.
  *
  * @returns {[
- *   [
- *     {
+ *   {
+ *     mutation: {
  *       args: [
  *         {
  *           _cylcObject: string,
@@ -370,7 +370,7 @@ export function getIntrospectionQuery () {
  *       ]
  *     },
  *     requiresInfo: boolean
- *   ]
+ *   }
  * ]} [satisfied, all]
  */
 export function filterAssociations (cylcObject, tokens, mutations) {
@@ -407,7 +407,7 @@ export function filterAssociations (cylcObject, tokens, mutations) {
       continue
     }
     ret.push(
-      [mutation, requiresInfo]
+      { mutation, requiresInfo } // expands automatically to { mutation: mutation, requiresInfo: requiresInfo }
     )
   }
   return ret
```

Then of course the user/caller of that function needs to change from `const [ mutation, requiresInfo ]  = ...` to either `ret = function(); mutation = ret.mutation ; requiresInfo = ret.requiresInfo` or `const { mutation, requiresInfo } = funcion()`.

But we don't need to change it here, it's easy to understand as it's now. Plus, there are other functions in Cylc UI that have the same problem, that's why I created #597, to see if we can add it to our GH actions later :+1:

[the remark this answers is not in the document]
Comment (kinow) [reply src/utils/aotf.js:372]: The "flavour" of jsdoc we are using in Cylc UI appears to be jsdoc + google-closure style. VScode appears to have another flavour of jsdoc where I think what you wrote might be valid. I'll look at #597 how to have either or both flavours in our CI :+1:

**Review by kinow — APPROVED** [review]

**Review by hjoliver — APPROVED** [review]
LGTM, had a good play with it :+1:

**Pull request merged**